src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingBackgroundService.cs

```diff
@@ -0,0 +1,487 @@
+        {
+            _processorId = Guid.NewGuid().ToString(); // consider using the pod name when running in AKS
+            _ownedPartitions = new List<int>();
+            _partitionLockingOptions = partitionLockingOptions;
+            _eventHubClientOptions = eventHubClientOptions;
+            _eventBatchingService = eventBatchingService;
+            _checkpointClient = checkpointClient;
+            _logger = logger;
+        }
+
+        private void CreateContainerClient()
+        {
+            var tokenCredential = EnsureArg.IsNotNull(_partitionLockingOptions.StorageTokenCredential, nameof(_partitionLockingOptions.StorageTokenCredential));
+            _containerClient = new BlobContainerClient(_partitionLockingOptions.BlobContainerUri, tokenCredential.GetCredential());
+            _containerClient.CreateIfNotExists();
+        }
+
+        private async Task RegisterProcessor(CancellationToken ct)
+        {
+            // write process id to storage
+            _logger.LogTrace($"Processor Id: {_processorId}");
+            BlobClient blobClient = _containerClient.GetBlobClient($"activepods/{_processorId}");
+
+            var activeUntil = new Dictionary<string, string>()
+            {
+                { "activeuntil", DateTime.UtcNow.AddMinutes(1).ToString() },
+            };
+
+            BlobUploadOptions options = new BlobUploadOptions() { Metadata = activeUntil };
+            await blobClient.UploadAsync(BinaryData.FromString(string.Empty), options, ct);
+
+            _logger.LogTrace($"Registered processor {_processorId}");
+        }
+
+        private async Task RegisterProcessorLoop(CancellationToken ct)
+        {
+            while (!ct.IsCancellationRequested)
+            {
+                await RegisterProcessor(ct);
+                await Task.Delay(15000, ct);
+            }
+        }
+
+        private async Task<string[]> GetPartitionsIdsForEventHub(CancellationToken ct)
+        {
+            await using (var consumer = new EventHubConsumerClient(
+                _eventHubClientOptions.EventHubConsumerGroup,
+                _eventHubClientOptions.EventHubNamespaceFQDN,
+                _eventHubClientOptions.EventHubName,
+                _eventHubClientOptions.EventHubTokenCredential))
+            {
+                return await consumer.GetPartitionIdsAsync(ct);
+            }
+        }
+
+        private async Task AcquirePartitions(CancellationToken ct)
+        {
+            // get all the processors in active status
+            var activeProcessors = ListBlobs(_containerClient, "activepods");
+            var activeProcessorCount = await GetActiveProcessors(activeProcessors, _containerClient, ct);
+            _totalProcessorsRunning = activeProcessorCount;
+            _logger.LogTrace($"Active processor count: {activeProcessorCount}");
+
+            if (activeProcessorCount == 0)
+            {
+                throw new Exception("active processors is 0");
```

Comment:
Should be a strongly typed/specific exception.

Reply:
updated to throw a typed exception